backend/danswer/indexing/chunker.py

```diff
@@ -122,192 +72,247 @@ def _get_metadata_suffix_for_document_index(
+    def _chunk_document(
+        self,
+        document: Document,
+        title_prefix: str,
+        metadata_suffix_semantic: str,
+        metadata_suffix_keyword: str,
+    ) -> list[DocAwareChunk]:
+        chunks: list[DocAwareChunk] = []
+        link_offsets: dict[int, str] = {}
+        chunk_text = ""
+        for section in document.sections:
+            section_text = section.text
+            section_link_text = section.link or ""
+
+            section_token_length = len(self.tokenizer.tokenize(section_text))
+            current_token_length = len(self.tokenizer.tokenize(chunk_text))
+            curr_offset_len = len(shared_precompare_cleanup(chunk_text))
+
+            # Large sections are considered self-contained/unique therefore they start a new chunk and are not concatenated
+            # at the end by other sections
+            if section_token_length > self.chunk_token_limit:
+                if chunk_text:
+                    chunks.append(
+                        DocAwareChunk(
+                            source_document=document,
+                            chunk_id=len(chunks),
+                            blurb=self._extract_blurb(chunk_text),
+                            content=chunk_text,
+                            source_links=link_offsets,
+                            section_continuation=False,
+                            title_prefix=title_prefix,
+                            metadata_suffix_semantic=metadata_suffix_semantic,
+                            metadata_suffix_keyword=metadata_suffix_keyword,
+                            mini_chunk_texts=self._get_mini_chunk_texts(chunk_text),
+                        )
+                    )
+                    link_offsets = {}
+                    chunk_text = ""
+
+                large_section_chunks = self._chunk_large_section(
+                    section_text=section_text,
+                    section_link_text=section_link_text,
+                    document=document,
+                    start_chunk_id=len(chunks),
+                    blurb=self._extract_blurb(section_text),
+                    title_prefix=title_prefix,
+                    metadata_suffix_semantic=metadata_suffix_semantic,
+                    metadata_suffix_keyword=metadata_suffix_keyword,
+                )
+                chunks.extend(large_section_chunks)
+                continue
+
+            # In the case where the whole section is shorter than a chunk, either adding to chunk or start a new one
+            if (
+                current_token_length
+                + len(self.tokenizer.tokenize(SECTION_SEPARATOR))
+                + section_token_length
+                <= self.chunk_token_limit
+            ):
+                chunk_text += SECTION_SEPARATOR * bool(chunk_text) + section_text
+                link_offsets[curr_offset_len] = section_link_text
+            else:
                 chunks.append(
                     DocAwareChunk(
```

Comment:
The code for creating chunks seems to share several parameters, so perhaps a new Factory class that takes those parameters once and avoids us having to pass them around repeatedly could help.

The rest of the parameters seem to be quite different in various places, so it raises concerns that perhaps we've jammed too many parameters into DocAwareChunk.  It's possible the different usages are really different objects that need to be initialized differently but share a similar chunk interface.